### DISCUSSION TITLE
How to get `tokio::net::TcpStream` out of `hyper::upgraded::Upgraded`?

### DISCUSSION THREAD

**Discussion by ncihnegn** [hyperium/hyper#3473] [Q&A]
Want to use the raw `TcpStream`.

**Comment by seanmonstar** [comment]
You can make use of [`Upgraded::downcast()`](https://docs.rs/hyper/latest/hyper/upgrade/struct.Upgraded.html#method.downcast).

**Comment by ncihnegn** [comment]
```
error[E0277]: the trait bound `tokio::net::TcpStream: hyper::rt::Read` is not satisfied
   --> src/main.rs:18:42
    |
18  |     let mut stream = upgraded.downcast::<TcpStream>();
    |                               --------   ^^^^^^^^^ the trait `hyper::rt::Read` is not implemented for `tokio::net::TcpStream`
    |                               |
    |                               required by a bound introduced by this call
    |
    = help: the following other types implement trait `hyper::rt::Read`:
              TokioIo<T>
              Box<T>
              Upgraded
              Pin<P>
              &mut T
note: required by a bound in `Upgraded::downcast`
```

**Reply by seanmonstar** [reply]
You need to use the same type as that wich you used for the `handshake(io)`/`serve_connection(io)`. Did you wrap it in a `TokioIo`?

**Reply by ncihnegn** [reply]
Yes. But I can't use `handshake(TcpStream)`.
```
error[E0277]: the trait bound `tokio::net::TcpStream: hyper::rt::Read` is not satisfied
   --> src/main.rs:41:68
    |
41  | ... conn) = hyper::client::conn::http1::handshake(stream).await?;
    |             ------------------------------------- ^^^^^^ the trait `hyper::rt::Read` is not implemented for `tokio::net::TcpStream`
```

**Reply by kitty-eu-org** [reply]
@ncihnegn Have you solved this problem?

**Reply by ncihnegn** [reply]
Not really, but using `tokio::io::copy_bidirectional` is good enough for me.

**Comment by ncihnegn** [comment]
Or how can I separate the reader and writer like `TcpStream::into_split()`?

**Comment by maxim-arcsec** [comment]
I ran into this and couldn't get [`Upgraded::downcast<T>`](https://docs.rs/hyper/latest/hyper/upgrade/struct.Upgraded.html#method.downcast) to work. Turns out the upgraded connection is wrapped inside a [`Rewind<T>`](https://github.com/hyperium/hyper-util/blob/e1c5a6c89bfaed11fb34bd483fe9ba616f403791/src/common/rewind.rs#L13-L16) which is private so it's impossible to provide the correct `T` for `Upgraded::downcast<T>` to succeed.

Instead you need to use the [`downcast`](https://docs.rs/hyper-util/latest/hyper_util/server/conn/auto/upgrade/fn.downcast.html) function from `hyper_util`.

```rust
use tokio::net::TcpStream;
use hyper_util::TokioIo;
use hyper_util::server::conn::auto::upgrade::downcast;

let tokio_io_stream: TokioIo<TcpStream> = downcast(upgraded).unwrap().io;
let tcp_stream: TcpStream = tokio_io_stream.into_inner();
```

This issue is mentioned [here](https://github.com/hyperium/hyper/issues/3704)
